### PR TITLE
Add missing type definitions

### PR DESCRIPTION
**Pull request by fireridlle**
### Notes
[DRV-435](https://faunadb.atlassian.net/browse/DRV-435)

Linked issues
https://github.com/fauna/faunadb-js/issues/384
https://github.com/fauna/faunadb-js/issues/341

### How to test
tsconfig.json should have "lib": ['es2019']. in that case ts will not inject type for `fetch` (will only if lib has `DOM`)
No typescript error should be thrown
```javascript

import { Client, errors, FaunaHttpErrorResponseContent, query, values } from './index';

const client = new Client({
    secret: '',
    observer: (res => {
        if((res.responseContent as FaunaHttpErrorResponseContent).errors) {
            console.error('observer receive error')
            console.error(res.responseContent)
        }
    })
})


client.query<values.FaunaTime>(query.Now())
.then(now => {
    console.info('now ', now.date)
})
.catch((err: errors.FaunaHTTPError) => {
    console.error('query catch error')
    console.error(err.requestResult.responseContent.errors)
})

```

### CODE REVIEW

**Review by parkhomenko — APPROVED** [review]

**Review by erickpintor — DISMISSED** [review]
Neat! 👍

File: src/types/Client.d.ts [hunk 1]
```diff
@@ -26,7 +28,7 @@ export interface QueryOptions {
 
 export default class Client {
   constructor(opts?: ClientConfig)
-  query<T = object>(expr: ExprArg, options?: QueryOptions): Promise<T>
+  query<T extends object = object>(expr: ExprArg, options?: QueryOptions): Promise<T>
```
Comment (tigger9flow) [inline src/types/Client.d.ts:31]: Fauna query result is not limited to type `object` only, e.g. for the query below it returns just a plain number:
```
.query(Q.Add(2, 2)) // Promise<number>
```
Another thing is it'd be great to use the `unknown` type as a default for any generic type in Fauna. But not sure whether it's backward-compatible :thinking_face:

File: src/RequestResult.js [hunk 2]
```diff
@@ -94,3 +94,19 @@ Object.defineProperty(RequestResult.prototype, 'timeTaken', {
 })
 
 module.exports = RequestResult
+
```
Comment (tigger9flow) [inline src/RequestResult.js:97]: Is it redundant?

File: src/types/Client.d.ts [hunk 3]
```diff
@@ -13,7 +15,7 @@ export interface ClientConfig {
   port?: number
   timeout?: number
   queryTimeout?: number
-  observer?: (res: RequestResult, client: Client) => void
+  observer?: <T extends object = object>(res: RequestResult<T | errors.FaunaHTTPError>, client: Client) => void
```
Comment (tigger9flow) [inline src/types/Client.d.ts:18]: Same for `object` type restriction as mentioned below

Comment (fireridlle) [reply src/types/Client.d.ts:18]: not really. observer accepts the full response or error

File: src/types/Client.d.ts (hunk 1 above)
Comment (fireridlle) [reply src/types/Client.d.ts]: removed `extends`. not sure about the `unknown` type. probably we will think about that no sooner than migrate driver code to typescript. so we can use that types during development and check how things work

File: src/RequestResult.js (hunk 2 above)
Comment (fireridlle) [reply src/RequestResult.js]: yes

**Review by parkhomenko — APPROVED** [review]

**Review by tigger9flow — APPROVED** [review]

**Pull request merged**